### ISSUE TITLE
Why strip stamps?

### ISSUE THREAD

**Issue by BerkeleyTrue** [stampit-org/react-stampit#21]
Hello All! 

I recently became a convert from es6 classes to stamps and loving every minute of it. Just converted a project of mine that had a bunch of es6 classes into stamps. It helped clean up my api in a great way!

I have a related project that uses es6 React class and I wanted to give this project a try. One component needs childContext working so react-stampit was not useable there, but createClass worked fine for that. The other has a lot of logic going on in the constructor method which I thought would work out fine as I can just use the stampit init method for that. Turns out all the beloved stampit methods are stripped from the factory.

What was the reasoning here? It seems to deflate the awesomeness that is the stamp.

**Comment by troutowicz** [comment]
The goal of react-stampit is to enforce an API familiar to React users while providing the power of stamp composition for component mixability. This means exposing only a subset of the stampit API.

> The other has a lot of logic going on in the constructor method

Can you provide an example of this logic? Are you referring to method binding boilerplate? I'm not sure what other logic would make sense in a React component. An `init` property may be a good idea nonetheless. :)

**Comment by JosephClay** [comment]
> The goal of react-stampit is to enforce an API familiar to React users while providing the power of stamp composition for component mixability.

I ran into similar surprises as @BerkeleyTrue when first picking up `react-stampit`...I had assumed that it was simply a stamp that was React-aware, but it seems there are more differences than that. 

Can we document the deviations from `stampit`, maybe with a short explanation of why for new-comers?

**Comment by troutowicz** [comment]
> Can we document the deviations from stampit

Is this not what the API spec in the README accomplishes?

**Comment by BerkeleyTrue** [comment]
Here is the script I am trying to convert https://github.com/ThunderCatsJS/thundercats-react/blob/master/src/createContainer.js and my attempt https://github.com/ThunderCatsJS/thundercats-react/blob/ofStampsAndMen/src/createContainer.js

I figured I could use compose and a regular stamp for the init function. I stopped on failing tests. I will have to come back to it later.

**Comment by BerkeleyTrue** [comment]
Api spec states what rStampit brings to the table. It does not explicitly state that stampit methods that a stamp should inherit are removed.

My assumption, and I am guessing the assumption of many others, is that the stamp would still have stampit methods and that they where not included in the docs as to not duplicate effort. This is similar to what I did [here](https://github.com/r3dm/thundercats), where I state that factories are stamps.

I think an additional note would help clarify this situation. But, I do think the whole stampit api should be exposed as they are what makes stamps powerful. Without it I don't see a huge advantage over createClass.

**Comment by JosephClay** [comment]
I understood the README in the same way as @BerkeleyTrue and the first several sections reinforced my misunderstanding. It's not until you hit **API** that there's a specific API laid out. Even then, it's easy to read it as an explanation and assume all the power of `stampit` still exists in the same way that `.addClass()` isn't lost when using a jquery plugin, or `.map()` when using a lodash mixin.

**Comment by troutowicz** [comment]
@BerkeleyTrue 

I think you might be missing the point of this library. If you want to use stampit, by all means use stampit. If you want an API designed specifically for React components, use react-stampit. There are many advantages to using react-stampit over createClass.
- no magical autobinding
- mixins are decoupled from a component's declaration
- React statics are declared with objects, not through functions returning objects

There may be other parts of the stampit API that could benefit react-stampit, I think an `init` property may make sense. But the entire stampit API will never be exposed with react-stampit.

@BerkeleyTrue @JosephClay 

I can understand the misinterpretation of the README. A PR clearing up the confusion would be awesome.

**Comment by JosephClay** [comment]
@troutowicz I will tackle a PR today or tomorrow, time permitting.

**Comment by troutowicz** [comment]
Thanks @JosephClay! :+1:

**Comment by BerkeleyTrue** [comment]
@troutowicz Fair enough. I definitely need `init`, though. 

I would also vote to include a react-safe `static` method similar to`stampit().static` that uses `extractStatics` underneath the hood.

**Comment by troutowicz** [comment]
I don't see a problem with exposing `init`. `statics` is available already as an object property. :)

cc @ericelliott

**Comment by BerkeleyTrue** [comment]
I am aware that there is a statics property and I'm not a fan of that. Not a fan of the statics property in react.createClass, either, and I would rather use a static method if thats possible.

**Comment by ericelliott** [comment]
> Is this not what the API spec in the README accomplishes?

Differences should be explicitly pointed out.

## The Future

Eventually, I'd like to see react-stampit turn into a standard [composable](https://github.com/stampit-org/stampit/issues/151).

Hopefully, following a standard specification, all of the future Stampit utilities should work with react-stampit. =)

**Comment by troutowicz** [comment]
After some thinking, I have changed my mind about exposing the `init` property.

>  I definitely need init, though.

@BerkeleyTrue, you really don't. The class constructor is just an alternative to using the React `componentWillMount` lifecycle method.

If you plan on continuing with implementing react-stampit, I'd like to suggest that for your HOC, you move the flux init logic into a mixin, and use some stamp composition. :)

**Comment by BerkeleyTrue** [comment]
Yup, thats what I ended up doing but I do not like the api. I will not pursue it further.

**Comment by troutowicz** [comment]
react-stampit's API was designed to be distantly similar to `React.createClass` and to be nothing like the ES6 `class` way of designing React components. If you prefer `class` then by all means you should use it! :)

If you are sticking with `class` you may be interested in the stamp decorator described [here](https://github.com/stampit-org/react-stampit/blob/master/docs/advanced.md).

**Comment by BerkeleyTrue** [comment]
I'm thinking in terms of audience targets for this project. I am willing to bet the majority of users coming to this project are those coming from stampit land.

They are going to expect, as I did, to be able to use stampit methods to create component factories. 

They are going to be those, like I was, that become disenfranchised with es6 classes and want something much more powerful then `createClass`.

I don't think the selling points you gave above are enough to justify using this package over createClass.

> no magical autobinding

Those coming from es6 classes don't have this already.

> mixins are decoupled from a component's declaration

This is nice but not a huge selling point.

> React statics are declared with objects, not through functions returning objects

Not sure what you mean by this as statics declaration seems to be the same as `createClass` so some clarification is probably necessary.

**Comment by troutowicz** [comment]
As I said before, you are still free to use core stampit to create your React components. This library must satisfy both long time stampit users as well as React users who have no idea what a stamp is.

> Those coming from es6 classes don't have this already.

I thought we were comparing react-stampit and createClass? 

> This is nice but not a huge selling point.

I disagree. One sets limits on component design, the other encourages composition.

> Not sure what you mean by this as statics declaration seems to be the same as createClass so some clarification is probably necessary.

``` js
React.createClass({
  getDefaultProps() {
    return { ... }
  }
});

stampit(React, {
  defaultProps: { ... }
});
```

That last point is purely personal taste.

**Comment by troutowicz** [comment]
Reopening this as it seems there is more to discuss.

**Comment by ericelliott** [comment]
> I'm thinking in terms of audience targets for this project. I am willing to bet the majority of users coming to this project are those coming from stampit land.

I'm not so sure. There are a lot of people in the React world looking for viable alternatives to `.createClass()` & ES6 `class`. I would guess that would be a very strong funnel to draw users to `react-stampit`.

> I don't think the selling points you gave above are enough to justify using this package over createClass.

What selling points would you focus on?

@troutowicz 

> I thought we were comparing react-stampit and createClass?

I think we need to contrast this with both `createClass` and ES6 `class`.

> One way sets limits on component design, the other encourages composition.

I agree, but we clearly need to expand on that in the documentation somehow.

**Comment by troutowicz** [comment]
> I think we need to contrast this with both createClass and ES6 class.

Oh I agree. @BerkeleyTrue's initial argument only concerned createClass, I was just staying on point.

**Comment by ericelliott** [comment]
Maybe we should explicitly list some of the disadvantages of `class`. See [How to Fix the ES6 `class` Keyword](https://medium.com/javascript-scene/how-to-fix-the-es6-class-keyword-2d42bb3f4caf) for some cons of class to document.

**Comment by troutowicz** [comment]
Closing, refer to #31
